examples/userapi/mpi.ipynb

```diff
@@ -0,0 +1,1521 @@
+      "completed": "2019-05-10T07:06:45.893666",
+      "data": {},
+      "engine_id": 3,
+      "engine_uuid": "11fe9a20-221ccab8fb1d468b901c320a",
+      "error": null,
+      "execute_input": "f.data",
+      "execute_result": {
+       "data": {
+        "text/plain": "Data([[3.75, 1.25],\n      [1.25, 0.  ]], dtype=float32)"
+       },
+       "execution_count": 12,
+       "metadata": {}
+      },
+      "follow": [],
+      "msg_id": "2fa91d31-2059e0dd23bdd4202044680e",
+      "outputs": [],
+      "received": "2019-05-10T07:06:45.898762",
+      "started": "2019-05-10T07:06:45.880628",
+      "status": "ok",
+      "stderr": "",
+      "stdout": "",
+      "submitted": "2019-05-10T07:06:45.872339"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "%%px\n",
+    "f.data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Performance optimizations\n",
+    "\n",
+    "The Devito compiler applies several optimizations before generating code.\n",
+    "\n",
+    "* Redundant halo exchanges are identified and removed. A halo exchange is redundant if a prior halo exchange carries out the same `Function` update and the data is not “dirty” yet.\n",
+    "* Computation/communication overlap, with explicit prodding of the asynchronous progress engine to make sure that non-blocking communications execute in background during the compute part.\n",
+    "* Halo exchanges could also be reshuffled to maximize the extension of the computation/communication overlap region.\n",
+    "\n",
+    "To run with all these optimizations enabled, instead of `DEVITO_MPI=1`, users should set `DEVITO_MPI=full`, or, equivalently"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%px\n",
+    "configuration['mpi'] = 'full'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We could now peek at the generated code to see that things now look differently."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "%%px\n",
+    "op = Operator(Eq(u.forward, u.dx + 1))\n",
+    "# Uncomment below to show code (it's quite verbose)\n",
+    "# print(op)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The body of the time-stepping loop has changed, as it now implements a classic computation/communication overlap scheme:\n",
+    "\n",
+    "* `haloupdate0` triggers non-blocking communications;\n",
+    "* `compute0` executes the core domain region, that is the sub-region which doesn't require reading from halo data to be computed;\n",
+    "* `halowait0` wait and terminates the non-blocking communications;\n",
+    "* `remainder0`, which internally calls `compute0`, computes the boundary region requiring the now up-to-date halo data."
```

Comment:
this is very clear.
maybe this pattern should be attended for talking about those functions by the first time (as above)

Reply:
good point